### PR TITLE
Fix cuSPARSE module build failure with CUDA 10.1 on Windows

### LINKED ISSUE #4412: Compilation Error from the latest Source Code in GitHub possibly caused by CUSPARESE APIs
Kindly need help. 
I'm currently trying to install cupy packages by compiling the source code released in the GitHub.

My System version is:
`Ubuntu 18.04`

My CUDA version is:
`Copyright (c) 2005-2019 NVIDIA Corporation`
`Built on Fri_Feb__8_19:08:17_PST_2019`
`Cuda compilation tools, release 10.1, V10.1.105`

My Python version is:
`Python 3.8.5`

And when installing the packages by the following commands:
`cd cupy`
`pip install . -vvvv`

Some errors occur:
`In file included from cupy_backends/cuda/libs/cusparse.cpp:676:0:`
`cupy_backends/cuda/libs/../cupy_cusparse.h:282:17: error: conflicting declaration ‘typedef enum cusparseIndexType_t cusparseIndexType_t’`
`typedef enum {} cusparseIndexType_t`
 `                             ^~~~~~~~~~~~~~~~~~~`
 ` In file included from cupy_backends/cuda/libs/../cupy_cusparse.h:7:0,
                   from cupy_backends/cuda/libs/cusparse.cpp:676:`
  `/usr/local/cuda-10.1/include/cusparse.h:6960:3: note: previous declaration as ‘typedef enum cusparseIndexType_t cusparseIndexType_t’
   } cusparseIndexType_t;
     ^~~~~~~~~~~~~~~~~~~`
 ` In file included from cupy_backends/cuda/libs/cusparse.cpp:676:0:
  cupy_backends/cuda/libs/../cupy_cusparse.h:283:17: error: conflicting declaration ‘typedef enum cusparseFormat_t cusparseFormat_t’
   typedef enum {} cusparseFormat_t;
                   ^~~~~~~~~~~~~~~~`
 ` In file included from cupy_backends/cuda/libs/../cupy_cusparse.h:7:0,
                   from cupy_backends/cuda/libs/cusparse.cpp:676:
  /usr/local/cuda-10.1/include/cusparse.h:6944:3: note: previous declaration as ‘typedef enum cusparseFormat_t cusparseFormat_t’
   } cusparseFormat_t;
     ^~~~~~~~~~~~~~~~`
 ` In file included from cupy_backends/cuda/libs/cusparse.cpp:676:0:
  cupy_backends/cuda/libs/../cupy_cusparse.h:284:17: error: conflicting declaration ‘typedef enum cusparseOrder_t cusparseOrder_t’
   typedef enum {} cusparseOrder_t;
                   ^~~~~~~~~~~~~~~`
  `In file included from cupy_backends/cuda/libs/../cupy_cusparse.h:7:0,
                   from cupy_backends/cuda/libs/cusparse.cpp:676:
  /usr/local/cuda-10.1/include/cusparse.h:6949:3: note: previous declaration as ‘typedef enum cusparseOrder_t cusparseOrder_t’
   } cusparseOrder_t;
     ^~~~~~~~~~~~~~~`
  `In file included from cupy_backends/cuda/libs/cusparse.cpp:676:0:
  cupy_backends/cuda/libs/../cupy_cusparse.h:286:17: error: conflicting declaration ‘typedef enum cusparseSpMMAlg_t cusparseSpMMAlg_t’
   typedef enum {} cusparseSpMMAlg_t;
                   ^~~~~~~~~~~~~~~~~`
`  In file included from cupy_backends/cuda/libs/../cupy_cusparse.h:7:0,
                   from cupy_backends/cuda/libs/cusparse.cpp:676:
  /usr/local/cuda-10.1/include/cusparse.h:6955:3: note: previous declaration as ‘typedef enum cusparseSpMMAlg_t cusparseSpMMAlg_t’
   } cusparseSpMMAlg_t;
     ^~~~~~~~~~~~~~~~~`
`  In file included from cupy_backends/cuda/libs/cusparse.cpp:676:0:
  cupy_backends/cuda/libs/../cupy_cusparse.h:450:17: error: conflicting declaration ‘typedef enum cusparseCsr2CscAlg_t cusparseCsr2CscAlg_t’
   typedef enum {} cusparseCsr2CscAlg_t;
                   ^~~~~~~~~~~~~~~~~~~~`
`  In file included from cupy_backends/cuda/libs/../cupy_cusparse.h:7:0,
                   from cupy_backends/cuda/libs/cusparse.cpp:676:
  /usr/local/cuda-10.1/include/cusparse.h:6898:3: note: previous declaration as ‘typedef enum cusparseCsr2CscAlg_t cusparseCsr2CscAlg_t’
   } cusparseCsr2CscAlg_t;
     ^~~~~~~~~~~~~~~~~~~~`

I search for several ways attempting to solve the conflicts, including finding the declarations in the `cusparse.h` included in CUDA-10,1 and trying to make correspondences between cupy's and CUDA's conflicted declarations. However, no one works. So I am sincerely asking for some help.

By the way, I have already tried to install by `pip install --pre cupy-cuda101`. There are also some errors in the linking reference of cupy to 'libcusparse.so', with the same familiar functions. I suspect there may be relationships between the compilation problems and the linking problems.

### PR DESCRIPTION
**Pull request by kmaehashi**
Fixes #4412

### CODE REVIEW

**Comment by kmaehashi** [comment]
Confirmed this worked with every CUDA 10.1 variant on Linux. Testing Windows now...

Jenkins, test this please.

**Comment by suntaochun** [comment]
Hi, is there any newly released updated version for the repairing of this bug on Linux? Experimental one is ok. 
@kmaehashi Thanks.

**Comment by leofang** [comment]
This will likely also fix https://github.com/conda-forge/cupy-feedstock/issues/78.

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by kmaehashi** [comment]
I confirmed this PR fixes the build issues with CUDA 10.1 / 10.1 Update 1 / 10.1 Update 2 on Windows.
pfnCI, test this please.

**Comment by kmaehashi** [comment]
For record:
Screen Shot 2021-02-17 at 11 40 21

File: cupy_backends/cuda/cupy_cusparse.h [hunk 1]
```diff
@@ -5,201 +5,440 @@
...
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpVecGetIndexBase(...) {
+cusparseStatus_t cusparseDestroySpVec(cusparseSpVecDescr_t spVecDescr) {
   return CUSPARSE_STATUS_SUCCESS;
 }
 
-cusparseStatus_t cusparseSpVecGetValues(...) {
+cusparseStatus_t cusparseSpVecGet(
```
Comment (emcastillo) [inline cupy_backends/cuda/cupy_cusparse.h:109]: nit: would be useful to have a comment here saying that the args were added to fix an issue in windows.

Comment (emcastillo) [reply cupy_backends/cuda/cupy_cusparse.h:109]: just a ref to the PR should be enough

Comment (kmaehashi) [reply cupy_backends/cuda/cupy_cusparse.h:114]: Thanks! Updated comments.

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Review by emcastillo — APPROVED** [review]

**Comment by kmaehashi** [comment]
pfnCI, test this please.

**Comment by suntaochun** [comment]
@kmaehashi Could you please allow me to be one of the contributors to the cuda_cusparse.h file? I thank and fully appreciate if possible while sincerely understand if not.

**Pull request merged**